feat(lsp): hide triple - in hover window

elianiva opened the pull request:
This thing has been bothering me for a while. Hover result can contain `---` because it's used as a horizontal line for markdown (similar to HTML's `<hr />`). Because Neovim can't display a thin horizontal line like other GUI text editor, I think removing it is a better solution. Here's some example from dart analyzer and sumneko lua.
Not sure why Dart's empty line is a bit different though.

| before | after |
| ---       | ---    |
|  ![Shot-2021-05-03_1](https://user-images.githubusercontent.com/51877647/116831511-1183c280-abda-11eb-805b-b5684173f540.png) | ![Shot-2021-05-03](https://user-images.githubusercontent.com/51877647/116831507-0892f100-abda-11eb-881e-28519727c92a.png) |
| ![Shot-2021-05-03_2](https://user-images.githubusercontent.com/51877647/116831525-2bbda080-abda-11eb-84eb-baa72d098847.png) | ![Shot-2021-05-03_3](https://user-images.githubusercontent.com/51877647/116831526-311aeb00-abda-11eb-8ba7-3ac0d5d05170.png) |

mjlbach commented:
Another option would be to add a conceal rule in `runtime/syntax/lsp_markdown.vim`

elianiva commented:
> Another option would be to add a conceal rule in runtime/syntax/lsp_markdown.vim

Ah, didn't thought of that. I think it would be better.

elianiva commented:
hmm, I'm not sure why it fails for some lines.

- incorrect
  ![image](https://user-images.githubusercontent.com/51877647/116835213-bc50ac80-abeb-11eb-874c-d81a08410403.png)

- correct
  ![image](https://user-images.githubusercontent.com/51877647/116835235-cd012280-abeb-11eb-9a5a-cda4ea99d1bd.png)

I also added a rule for an Identifier which looks like `[mainAxisAlignment]` to hide the bracket and link them to the `Identifier` highlight group.

teto commented:
I dont think we should remove it, what about GUIs that can interpret it ? there are also some tools that can render markdown in a TUI so one could pipe the output in it and couple it with nvim-terminal to interpret the terminal codes.

elianiva commented:
yes, I'm going with conceal instead.

elianiva commented:
okay, so the issue here is `markdownH2` always takes precedence if the hover content looks like this:

```lua
local modules: {}
─────────────────
--- <- this does not get concealed
 some random description idk
```

It can only match if there's an empty line above the `---`, like this:

```lua
local modules: {}
─────────────────

--- <- this gets concealed
 some random description idk
```

mjlbach commented:
@elianiva Did you end up figuring it out?

elianiva commented:
nope, I also found out that in some cases, the border for code snippet and description has different highlight.
Sometimes it's captured as `htmlH1` but sometimes it doesn't.

elianiva committed:
feat(lsp): hide triple - in hover window.

elianiva commented:
I managed to solve my previous issue, now they're all hidden. I removed the rule for `markdownH2` which was causing the issue, I haven't notice any regression from doing this so if anyone want to review this, that would be nice.

My previous issue was this, it doesn't get hidden even if I added a rule to hide them.
![image](https://user-images.githubusercontent.com/51877647/121440181-9f10ba00-c9b1-11eb-94a9-0c74cb80c994.png)

Before: 
![image](https://user-images.githubusercontent.com/51877647/121440149-8f917100-c9b1-11eb-8a0b-8f981d012749.png)

After:
![image](https://user-images.githubusercontent.com/51877647/121440067-5fe26900-c9b1-11eb-9fa3-0b502f7381f9.png)
![image](https://user-images.githubusercontent.com/51877647/121440131-86080900-c9b1-11eb-8752-e9fd505d55df.png)
